### ISSUE TITLE
An issue about for statement without body

### ISSUE THREAD

**Issue by YuanWangC** [chakra-core/ChakraCore#6351]
When running the testcase below, a SyntaxError exception should be thrown while chakra does not. 
According to ECMAScript Language Specification( from ES 5.1 to [ES 10](http://www.ecma-international.org/ecma-262/10.0/index.html#sec-for-statement) ), `for statement` should have a body with a type of `Statement`. This maybe an issue of Chakra.
#### version
[eaaf7ac](https://github.com/microsoft/ChakraCore/commit/eaaf7ac6beca01ab0d80c05959f3868463043c2d)
#### testcase
```
var NISLFuzzingFunc = function(){
    eval("for(;false;)");    
    print("RunHere");
};
NISLFuzzingFunc();
```
#### output
`RunHere`
#### expected output
` SyntaxError: Unexpected end of script`

**Comment by fatcerberus** [comment]
I first thought this was due to automatic semicolon insertion (ASI) parsing the statement as
```js
for (; false; ) ;
```
(a for-loop with an empty statement for a body), but Node throws a `SyntaxError` for the same code:
```
> eval("for(;false;)")
Thrown:
SyntaxError: Unexpected end of input
```
so it looks like you're correct that this should be an error.

**Comment by rhuanjl** [comment]
It's weird one because (per other engines) it is a syntax error inside eval OR if the next character is a } or a ] (or anything ending a statement) BUT it's not a syntax error outside of eval if the next character is the end of the script.

CC gets all cases right per comparison with other engines OTHER THAN the inside eval case when it's the end of the script you're eval'ing.

**Comment by rhuanjl** [comment]
This is a relatively easy fix - edit needed would be in parse.cpp This would be a great starting point for a first time contributor, if anyone wishes to pick this up I'll be happy to help as needed.

**Comment by vinaygudarad** [comment]
I would like to work on this. I am a first time contributor So I need u to walk me through what I should be doing like where to start and where the changes need to be done.

**Comment by rhuanjl** [comment]
@vinaygudarad great, not sure how much experience you have so this be too much or too little guidance:
1. Fork chakracore, checkout the master branch, and make a new branch from it
2. Make sure you can build chakracore locally (on windows you'll need msbuild installed, on linux or macos you'll need cmake installed) then on windows open a developer command prompt in the chakracore folder and run `jenkins\buildone.cmd x64 debug` on linux or macos run `./build.sh --debug --static --embed-icu -j=2`
3. Check it works (run `ch.exe` on windows or `ch` on macos/linux - the build will have created this app)
4. Now you have chakracore working you can make an edit - try running the provided test case above with ch you should get the incorrect output reported above
5. Now take a look in chakracore/parser/Parse.cpp at what happens when it's parsing a for loop - the relevant part starts on line 10299. You'll need to work it when it gets to the right bracket of the for condition and see what it does next - it specifically needs to error if it finds an end of input at that point.
6. Edit Parse.cpp and rebuild/retest untill you have the error fixed.
7. Add the test case, possibly as part of test/Bugs/misc_bugs.js (look at that file and copy how the cases are setup) - you'll need to use assert.throws so the test passes only if the eval throws.
8. Check what you've done hasn't broken anything by running the test suite offline: on windows `jenkins/testone.cmd x64 debug` on linux or macos `test/runtests.py -d`
8. Open a pull request

**Comment by vinaygudarad** [comment]
I am a beginner with basic cpp knowledge but I want to learn and work on
this thing. So please tell me what to read and if possible provide the link
to install msbuild.

**Comment by rhuanjl** [comment]
msbuild is a microsoft build tool it's installed alongside MS Visual Studio (including with the free version), not sure if you can get it's on its own. Have a look: https://visualstudio.microsoft.com/downloads/

Once you have chakracore checked out locally and building I can give you more pointers if you find the code hard to follow.

**Comment by vinaygudarad** [comment]
I am a beginner with basic cpp knowledge but I want to learn necessary
things and work on this thing. So if possible please provide necessary
links for applications to install and start working. Thank You :)

**Comment by ppenzin** [comment]
@vinaygudarad you would need Visual Studio (Community Edition at the least) - make sure C++ is selected in the installer. Visual Studio Community Edition would not be able to open the solution in the GUI, but it would be able to build it - don't be discouraged if you open the solution and get some form of "cannot load project" error - it should still build build using the command @rhuanjl provided above.

Also, if you are familiar with Unix-like OSes or want to try, you can use the free Ubuntu App from Windows store - that way you can edit files in "normal" Windows way, but then run Linux variant of the build via the app (which executes a Windows subsystem for Linux Virtual Machine). Let me know if you want to try that, I can give you a few pointers on setting up.

**Comment by vinaygudarad** [comment]
@ppenzin  am trying to install visual studio community edition 2019 through visual studio installer. I selected C++ in the installer. So thank you for your guidance and will ask you if i need anything once its installed.
![image](https://user-images.githubusercontent.com/46954468/82187453-13417100-990a-11ea-80a2-6400fc4a05fe.png) This is the development tool I am installing. If I am doing anything wrong, please let me know.

**Comment by rhuanjl** [comment]
Looks about right - I'd double check the list on the right to ensure msbuild is ticked if it's listed (it may be included as part of the MSVC build tools or may be it's own line).

If you want to save diskspace you could remove a few components, e.g. you don't need the "Just-in-Time debugger" or the test adapters or "Live Share".

One other note - once it's installed VS Community can actually open the chakracore solution file - I personally don't like working within it, I prefer to just build from the command line and use a different editor but you can do everything inside VS Community if you want to.

**Comment by ppenzin** [comment]
@vinaygudarad, I think sufficient version of MSBuild comes with any build tools bundle - you have that selected.

@rhuanjl were you able to build full solution with Visual Studio Community? When I tried to open it I got some form of "project did not load" message for one of sub-projects, which I thought might prevent it from building it successfully, though that should be still sufficient for editing files.

**Comment by rhuanjl** [comment]
@ppenzin That's weird - yes I could build it with VS Community.

**Comment by ppenzin** [comment]
Anyway, I checked - it builds just fine, so even if there any messages when loading the project, everything should just work.

**Comment by rhuanjl** [comment]
Note issue #6553 highlighted that if statements without bodies have the same issue as for statements without bodies - I closed it as a duplicate of this as the underlying point is the same but when fixing this should include a test case for if() as well.
